src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+    trace_msg(nullptr, "No vectorizable statement is found");
+    return false;
+  }
+
+  // Then, extend each statement by recursively including input nodes. Bail out
+  // if unsupported node is found.
+  for (int idx = 0; idx < num_stmts; idx++) {
+    Node_List* stmt = _stmts.at(idx);
+    assert(stmt->size() == 1, "Each statement should have exactly one node");
+    Node* store = stmt->at(0);
+    // Add value input of the store node into a worklist to include more nodes
+    // into current statement
+    Node_List* worklist = new Node_List(_arena);
+    if (!collect_statements_helper(store, MemNode::ValueIn, stmt, worklist)) {
+      return false;
+    }
+    // Continue adding nodes until the worklist is empty
+    while (worklist->size() > 0) {
+      Node* node = worklist->pop();
+      uint start, end;
+      VectorNode::vector_operands(node, &start, &end);
+      for (uint idx = start; idx < end; idx++) {
+        if (!collect_statements_helper(node, idx, stmt, worklist)) {
+          return false;
+        }
+      }
+    }
+#ifndef PRODUCT
+    if (TraceMaskedLoop) {
+      tty->print_cr("Nodes in statement [%d] with element type '%s'",
+                    idx, type2name(statement_bottom_type(stmt)));
+      for (uint i = 0; i < stmt->size(); i++) {
+        stmt->at(i)->dump();
+      }
+      tty->cr();
+    }
+#endif
+  }
+
+  return true;
+}
+
+// Analyze loop statements and bail out if any of them is not vectorizable
+bool VectorMaskedLoop::analyze_vectorizability() {
+  if (!find_vector_element_types()) {
+    return false;
+  }
+  if (!vector_nodes_implemented()) {
+    return false;
+  }
+  // Delegate data dependence check to SWPointer utility
+  if (SWPointer::has_potential_dependence(_swptrs)) {
+    trace_msg(nullptr, "Potential data dependence is found in the loop");
+    return false;
+  }
+  if (!analyze_loop_body_nodes()) {
+    return false;
+  }
+  return true;
+}
+
+// Find element basic type for each vectorization candidate node
+bool VectorMaskedLoop::find_vector_element_types() {
+  for (int idx = 0; idx < _stmts.length(); idx++) {
+    Node_List* stmt = _stmts.at(idx);
+    BasicType stmt_bottom_type = statement_bottom_type(stmt);
+    bool subword_stmt = is_subword_type(stmt_bottom_type);
+
+    // Record vector lane size
+    _size_stats.record_size(type2aelembytes(stmt_bottom_type));
+
+    // Set element type for each statement node from bottom to top. Bail out if
+    // the pattern is unsupported
+    for (int i = stmt->size() - 1; i >= 0; i--) {
+      Node* node = stmt->at(i);
+      if (node->is_Mem()) {
+        // Use memory type as its element basic type for memory node
+        BasicType mem_type = node->as_Mem()->memory_type();
+        set_elem_bt(node, mem_type);
+        if (node->is_Load()) {
+          // For load node, check if it has the same vector element size with
+          // the bottom type of the statement
+          if (!same_element_size(mem_type, stmt_bottom_type)) {
```

Comment:
Can this limitation be removed in the future?

Reply:
Yes, we have tried supporting type conversions (between different type sizes) but current solution is not mature and not included in this patch. So this limitation is added here.

Reply:
Ok, fine. Leave that for the future.

Reply:
I close this comment as the feature will not be included in this patch.